### PR TITLE
Implemented Dialog Types

### PR DESCRIPTION
**Pull request by Pokye**
* **Addressed Issue(s)**: https://github.com/rathena/rathena/issues/8109

* **Server Mode**: Pre-renewal and Renewal

* **Description of Pull Request**: 
Add script command for set npc dialog align: setdialogalign (20210203+).
Add script command for set npc dialog size: setdialogsize (20220504+).
Add script commands for set npc dialog position: setdialogpos, setdialogpospercent (20220504+).

Special thanks to @4144 and **#HerculesWS**

### CODE REVIEW

File: doc/script_commands.txt [hunk 1]
```diff
@@ -11412,3 +11412,37 @@ Example:
...
 ---------------------------------------
+
+*setdialogalign(<align>);
+
+Set vertical or horizontal align in npc dialog.
+Valid aligns:
+- horizontal align:
+  DIALOG_ALIGN_LEFT
+  DIALOG_ALIGN_RIGHT
+  DIALOG_ALIGN_CENTER
```
Comment (aleos89) [inline doc/script_commands.txt:11423]:
```suggestion
  DIALOG_ALIGN_CENTER
  DIALOG_ALIGN_RIGHT
```
Just to by the same as how the vertical align is described below.

File: doc/script_commands.txt [hunk 2]
```diff
@@ -11412,3 +11412,37 @@ Example:
     autoloot(10000); // 100.00%
 
 ---------------------------------------
+
+*setdialogalign(<align>);
+
+Set vertical or horizontal align in npc dialog.
```
Comment (aleos89) [inline doc/script_commands.txt:11418]:
```suggestion
Set vertical or horizontal align in NPC dialog.
```

File: doc/script_commands.txt [hunk 3]
```diff
@@ -11412,3 +11412,37 @@ Example:
...
+- vertical align:
+  DIALOG_ALIGN_TOP
+  DIALOG_ALIGN_MIDDLE
+  DIALOG_ALIGN_BOTTOM
+
+---------------------------------------
+
+*setdialogsize(<width>, <height>)
+
+Set size for npc dialog.
```
Comment (aleos89) [inline doc/script_commands.txt:11434]:
```suggestion
Set size for NPC dialog.
```

File: doc/script_commands.txt [hunk 4]
```diff
@@ -11412,3 +11412,37 @@ Example:
...
+
+---------------------------------------
+
+*setdialogpos(<x>, <y>)
+
+Set position for npc dialog in pixels.
```
Comment (aleos89) [inline doc/script_commands.txt:11440]:
```suggestion
Set position for NPC dialog in pixels.
```

File: doc/script_commands.txt [hunk 5]
```diff
@@ -11412,3 +11412,37 @@ Example:
...
+
+---------------------------------------
+
+*setdialogpospercent(<x>, <y>)
+
+Set position for npc dialog in screen size percent.
```
Comment (aleos89) [inline doc/script_commands.txt:11446]:
```suggestion
Set position for NPC dialog in screen size percent.
```

File: src/map/clif.cpp [hunk 6]
```diff
@@ -25187,6 +25187,65 @@ void clif_parse_reset_skill( int fd, map_session_data* sd ){
 #endif
 }
 
+void clif_sayDialogAlign(map_session_data *sd, int npcid, enum say_dialog_align align)
```
Comment (aleos89) [inline src/map/clif.cpp:25190]:
```suggestion
void clif_sayDialogAlign(map_session_data *sd, int npcid, e_say_dialog_align align)
```

File: src/map/clif.hpp [hunk 7]
```diff
@@ -1254,4 +1254,9 @@ void clif_macro_reporter_status(map_session_data &sd, e_macro_report_status styp
 
 void clif_dynamicnpc_result( map_session_data& sd, e_dynamicnpc_result result );
 
+void clif_sayDialogAlign(map_session_data *sd, int npcid, enum say_dialog_align align);
```
Comment (aleos89) [inline src/map/clif.hpp:1257]:
```suggestion
void clif_sayDialogAlign(map_session_data *sd, int npcid, e_say_dialog_align align);
```

File: src/map/map.hpp [hunk 8]
```diff
@@ -861,6 +861,18 @@ struct map_data_other_server {
 	uint16 port;
 };
 
+/**
+ * align for packet ZC_SAY_DIALOG_ALIGN
+ **/
+enum say_dialog_align {
```
Comment (aleos89) [inline src/map/map.hpp:867]:
```suggestion
enum e_say_dialog_align : uint8 {
```

File: src/map/script.cpp [hunk 9]
```diff
@@ -27123,6 +27123,54 @@ BUILDIN_FUNC(opentips){
...
 }
 
+BUILDIN_FUNC(setdialogalign){
+	map_session_data *sd;
+
+	if ( !script_rid2sd(sd) ) {
+		return SCRIPT_CMD_FAILURE;
+	}
+
+	clif_sayDialogAlign(sd, st->oid, (enum say_dialog_align)script_getnum(st, 2));
```
Comment (aleos89) [inline src/map/script.cpp:27133]:
```suggestion
	clif_sayDialogAlign(sd, st->oid, (e_say_dialog_align)script_getnum(st, 2));
```

**Review by aleos89 — APPROVED** [review]

File: src/map/clif.hpp [hunk 10]
```diff
@@ -1254,4 +1254,9 @@ void clif_macro_reporter_status(map_session_data &sd, e_macro_report_status styp
 
 void clif_dynamicnpc_result( map_session_data& sd, e_dynamicnpc_result result );
 
+void clif_sayDialogAlign(map_session_data *sd, int npcid, e_say_dialog_align align);
```
Comment (Lemongrass3110) [inline src/map/clif.hpp:1257]:
```suggestion
void clif_set_dialog_align(map_session_data& sd, int npcid, e_say_dialog_align align);
```

File: src/map/clif.hpp [hunk 11]
```diff
@@ -1254,4 +1254,9 @@ void clif_macro_reporter_status(map_session_data &sd, e_macro_report_status styp
...
+void clif_set_npc_window_size(map_session_data *sd, int width, int height);
```
Comment (Lemongrass3110) [inline src/map/clif.hpp:1258]:
```suggestion
void clif_set_npc_window_size(map_session_data& sd, int width, int height);
```

File: src/map/clif.hpp [hunk 12]
```diff
@@ -1254,4 +1254,9 @@ void clif_macro_reporter_status(map_session_data &sd, e_macro_report_status styp
...
+void clif_set_npc_window_pos(map_session_data *sd, int x, int y);
```
Comment (Lemongrass3110) [inline src/map/clif.hpp:1259]:
```suggestion
void clif_set_npc_window_pos(map_session_data& sd, int x, int y);
```

File: src/map/clif.hpp [hunk 13]
```diff
@@ -1254,4 +1254,9 @@ void clif_macro_reporter_status(map_session_data &sd, e_macro_report_status styp
...
+void clif_set_npc_window_pos_percent(map_session_data *sd, int x, int y);
```
Comment (Lemongrass3110) [inline src/map/clif.hpp:1260]:
```suggestion
void clif_set_npc_window_pos_percent(map_session_data& sd, int x, int y);
```

File: src/map/script.cpp [hunk 14]
```diff
@@ -27123,6 +27123,54 @@ BUILDIN_FUNC(opentips){
...
 }
 
+BUILDIN_FUNC(setdialogalign){
+	map_session_data *sd;
+
+	if ( !script_rid2sd(sd) ) {
+		return SCRIPT_CMD_FAILURE;
+	}
+
+	clif_sayDialogAlign(sd, st->oid, (e_say_dialog_align)script_getnum(st, 2));
```
Comment (Lemongrass3110) [inline src/map/script.cpp:27133]:
```suggestion
	clif_set_dialog_align(*sd, st->oid, (e_say_dialog_align)script_getnum(st, 2));
```

File: src/map/script.cpp [hunk 15]
```diff
@@ -27123,6 +27123,54 @@ BUILDIN_FUNC(opentips){
...
+
+BUILDIN_FUNC(setdialogsize){
+	map_session_data *sd;
+
+	if ( !script_rid2sd(sd) ) {
+		script_pushint(st, 0);
+		return SCRIPT_CMD_FAILURE;
+	}
+
+	clif_set_npc_window_size(sd, script_getnum(st, 2), script_getnum(st, 3));
```
Comment (Lemongrass3110) [inline src/map/script.cpp:27146]:
```suggestion
	clif_set_npc_window_size(*sd, script_getnum(st, 2), script_getnum(st, 3));
```

File: src/map/script.cpp [hunk 16]
```diff
@@ -27123,6 +27123,54 @@ BUILDIN_FUNC(opentips){
...
+
+BUILDIN_FUNC(setdialogpos){
+	map_session_data *sd;
+
+	if ( !script_rid2sd(sd) ) {
+		script_pushint(st, 0);
+		return SCRIPT_CMD_FAILURE;
+	}
+
+	clif_set_npc_window_pos(sd, script_getnum(st, 2), script_getnum(st, 3));
```
Comment (Lemongrass3110) [inline src/map/script.cpp:27158]:
```suggestion
	clif_set_npc_window_pos(*sd, script_getnum(st, 2), script_getnum(st, 3));
```

File: src/map/script.cpp [hunk 17]
```diff
@@ -27123,6 +27123,54 @@ BUILDIN_FUNC(opentips){
...
+
+BUILDIN_FUNC(setdialogpospercent){
+	map_session_data *sd;
+
+	if ( !script_rid2sd(sd) ) {
+		script_pushint(st, 0);
+		return SCRIPT_CMD_FAILURE;
+	}
+
+	clif_set_npc_window_pos_percent(sd, script_getnum(st, 2), script_getnum(st, 3));
```
Comment (Lemongrass3110) [inline src/map/script.cpp:27170]:
```suggestion
	clif_set_npc_window_pos_percent(*sd, script_getnum(st, 2), script_getnum(st, 3));
```

File: src/map/clif.cpp [hunk 18]
```diff
@@ -25187,6 +25187,65 @@ void clif_parse_reset_skill( int fd, map_session_data* sd ){
 #endif
 }
 
+void clif_sayDialogAlign(map_session_data *sd, int npcid, e_say_dialog_align align)
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:25190]:
```suggestion
void clif_set_dialog_align(map_session_data& sd, int npcid, e_say_dialog_align align)
```

File: src/map/clif.cpp [hunk 19]
```diff
@@ -25187,6 +25187,65 @@ void clif_parse_reset_skill( int fd, map_session_data* sd ){
...
+	const int fd = sd->fd;
+	WFIFOHEAD(fd, sizeof(struct PACKET_ZC_DIALOG_TEXT_ALIGN));
+	struct PACKET_ZC_DIALOG_TEXT_ALIGN *p = (struct PACKET_ZC_DIALOG_TEXT_ALIGN *)WFIFOP(fd, 0);
+	p->PacketType = HEADER_ZC_DIALOG_TEXT_ALIGN;
+	p->align = align;
+	WFIFOSET(fd, sizeof(struct PACKET_ZC_DIALOG_TEXT_ALIGN));
+#endif  // PACKETVER_MAIN_NUM >= 20210203 || PACKETVER_RE_NUM >= 20211103 || PACKETVER_ZERO_NUM >= 20221024
+}
+
+void clif_set_npc_window_size(map_session_data *sd, int width, int height)
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:25204]:
```suggestion
void clif_set_npc_window_size(map_session_data& sd, int width, int height)
```

File: src/map/clif.cpp [hunk 20]
```diff
@@ -25187,6 +25187,65 @@ void clif_parse_reset_skill( int fd, map_session_data* sd ){
...
+	WFIFOHEAD(fd, sizeof(struct PACKET_ZC_DIALOG_WINDOW_SIZE));
+	struct PACKET_ZC_DIALOG_WINDOW_SIZE *p = (struct PACKET_ZC_DIALOG_WINDOW_SIZE*)WFIFOP( fd, 0 );
+	p->PacketType = HEADER_ZC_DIALOG_WINDOW_SIZE;
+	p->width = width;
+	p->height = height;
+	WFIFOSET(fd, sizeof(struct PACKET_ZC_DIALOG_WINDOW_SIZE));
+#endif  // PACKETVER_MAIN_NUM >= 20220504
+}
+
+void clif_set_npc_window_pos(map_session_data *sd, int x, int y)
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:25219]:
```suggestion
void clif_set_npc_window_pos(map_session_data& sd, int x, int y)
```

File: src/map/clif.cpp [hunk 21]
```diff
@@ -25187,6 +25187,65 @@ void clif_parse_reset_skill( int fd, map_session_data* sd ){
...
+	WFIFOHEAD(fd, sizeof(struct PACKET_ZC_DIALOG_WINDOW_POS));
+	struct PACKET_ZC_DIALOG_WINDOW_POS *p = (struct PACKET_ZC_DIALOG_WINDOW_POS*)WFIFOP( fd, 0 );
+	p->PacketType = HEADER_ZC_DIALOG_WINDOW_POS;
+	p->x = x;
+	p->y = y;
+	WFIFOSET(fd, sizeof(struct PACKET_ZC_DIALOG_WINDOW_POS));
+#endif  // PACKETVER_MAIN_NUM >= 20220504
+}
+
+void clif_set_npc_window_pos_percent(map_session_data *sd, int x, int y)
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:25234]:
```suggestion
void clif_set_npc_window_pos_percent(map_session_data& sd, int x, int y)
```

File: src/map/clif.cpp [hunk 22]
```diff
@@ -25187,6 +25187,65 @@ void clif_parse_reset_skill( int fd, map_session_data* sd ){
...
+{
+#if PACKETVER_MAIN_NUM >= 20220504
+	nullpo_retv(sd);
+
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:25238]:
```suggestion
```
sd is a reference now and cannot be null.

File: src/map/clif.cpp [hunk 23]
```diff
@@ -25187,6 +25187,65 @@ void clif_parse_reset_skill( int fd, map_session_data* sd ){
...
+	p->height = height;
+	WFIFOSET(fd, sizeof(struct PACKET_ZC_DIALOG_WINDOW_SIZE));
+#endif  // PACKETVER_MAIN_NUM >= 20220504
+}
+
+void clif_set_npc_window_pos(map_session_data *sd, int x, int y)
+{
+#if PACKETVER_MAIN_NUM >= 20220504
+	nullpo_retv(sd);
+
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:25223]:
```suggestion
```
sd is a reference now and cannot be null.

File: src/map/clif.cpp [hunk 24]
```diff
@@ -25187,6 +25187,65 @@ void clif_parse_reset_skill( int fd, map_session_data* sd ){
...
+	p->align = align;
+	WFIFOSET(fd, sizeof(struct PACKET_ZC_DIALOG_TEXT_ALIGN));
+#endif  // PACKETVER_MAIN_NUM >= 20210203 || PACKETVER_RE_NUM >= 20211103 || PACKETVER_ZERO_NUM >= 20221024
+}
+
+void clif_set_npc_window_size(map_session_data *sd, int width, int height)
+{
+#if PACKETVER_MAIN_NUM >= 20220504
+	nullpo_retv(sd);
+
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:25208]:
```suggestion
```
sd is a reference now and cannot be null.

File: src/map/clif.cpp [hunk 25]
```diff
@@ -25187,6 +25187,65 @@ void clif_parse_reset_skill( int fd, map_session_data* sd ){
 #endif
 }
 
+void clif_sayDialogAlign(map_session_data *sd, int npcid, e_say_dialog_align align)
+{
+#if PACKETVER_MAIN_NUM >= 20210203 || PACKETVER_RE_NUM >= 20211103 || PACKETVER_ZERO_NUM >= 20221024
+	nullpo_retv(sd);
+
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:25194]:
```suggestion
```
sd is a reference now and cannot be null.

File: src/map/clif.cpp [hunk 26]
```diff
@@ -25187,6 +25187,65 @@ void clif_parse_reset_skill( int fd, map_session_data* sd ){
...
+#endif  // PACKETVER_MAIN_NUM >= 20210203 || PACKETVER_RE_NUM >= 20211103 || PACKETVER_ZERO_NUM >= 20221024
+}
+
+void clif_set_npc_window_size(map_session_data *sd, int width, int height)
+{
+#if PACKETVER_MAIN_NUM >= 20220504
+	nullpo_retv(sd);
+
+	const int fd = sd->fd;
+	WFIFOHEAD(fd, sizeof(struct PACKET_ZC_DIALOG_WINDOW_SIZE));
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:25210]:
```suggestion
```

File: src/map/clif.cpp [hunk 27]
```diff
@@ -25187,6 +25187,65 @@ void clif_parse_reset_skill( int fd, map_session_data* sd ){
...
+	struct PACKET_ZC_DIALOG_WINDOW_SIZE *p = (struct PACKET_ZC_DIALOG_WINDOW_SIZE*)WFIFOP( fd, 0 );
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:25211]:
```suggestion
	PACKET_ZC_DIALOG_WINDOW_SIZE p = {};

```

File: src/map/clif.cpp [hunk 28]
```diff
@@ -25187,6 +25187,65 @@ void clif_parse_reset_skill( int fd, map_session_data* sd ){
...
+	p->PacketType = HEADER_ZC_DIALOG_WINDOW_SIZE;
+	p->width = width;
+	p->height = height;
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:25214]:
```suggestion
	p.PacketType = HEADER_ZC_DIALOG_WINDOW_SIZE;
	p.width = width;
	p.height = height;
```

File: src/map/clif.cpp [hunk 29]
```diff
@@ -25187,6 +25187,65 @@ void clif_parse_reset_skill( int fd, map_session_data* sd ){
...
+	WFIFOSET(fd, sizeof(struct PACKET_ZC_DIALOG_WINDOW_SIZE));
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:25215]:
```suggestion

	clif_send( &p, sizeof( p ), &sd.bl, SELF );
```

File: src/map/clif.cpp [hunk 30]
```diff
@@ -25187,6 +25187,65 @@ void clif_parse_reset_skill( int fd, map_session_data* sd ){
...
+{
+#if PACKETVER_MAIN_NUM >= 20210203 || PACKETVER_RE_NUM >= 20211103 || PACKETVER_ZERO_NUM >= 20221024
+	nullpo_retv(sd);
+
+	const int fd = sd->fd;
+	WFIFOHEAD(fd, sizeof(struct PACKET_ZC_DIALOG_TEXT_ALIGN));
+	struct PACKET_ZC_DIALOG_TEXT_ALIGN *p = (struct PACKET_ZC_DIALOG_TEXT_ALIGN *)WFIFOP(fd, 0);
+	p->PacketType = HEADER_ZC_DIALOG_TEXT_ALIGN;
+	p->align = align;
+	WFIFOSET(fd, sizeof(struct PACKET_ZC_DIALOG_TEXT_ALIGN));
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:25200]:
```suggestion

	clif_send( &p, sizeof( p ), &sd.bl, SELF );
```

File: src/map/clif.cpp [hunk 31]
```diff
@@ -25187,6 +25187,65 @@ void clif_parse_reset_skill( int fd, map_session_data* sd ){
...
+#if PACKETVER_MAIN_NUM >= 20220504
+	nullpo_retv(sd);
+
+	const int fd = sd->fd;
+	WFIFOHEAD(fd, sizeof(struct PACKET_ZC_DIALOG_WINDOW_POS2));
+	struct PACKET_ZC_DIALOG_WINDOW_POS2 *p = (struct PACKET_ZC_DIALOG_WINDOW_POS2*)WFIFOP( fd, 0 );
+	p->PacketType = HEADER_ZC_DIALOG_WINDOW_POS2;
+	p->x = x;
+	p->y = y;
+	WFIFOSET(fd, sizeof(struct PACKET_ZC_DIALOG_WINDOW_POS2));
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:25245]:
```suggestion

	clif_send( &p, sizeof( p ), &sd.bl, SELF );
```

File: src/map/clif.cpp [hunk 32]
```diff
@@ -25187,6 +25187,65 @@ void clif_parse_reset_skill( int fd, map_session_data* sd ){
...
+#if PACKETVER_MAIN_NUM >= 20220504
+	nullpo_retv(sd);
+
+	const int fd = sd->fd;
+	WFIFOHEAD(fd, sizeof(struct PACKET_ZC_DIALOG_WINDOW_POS));
+	struct PACKET_ZC_DIALOG_WINDOW_POS *p = (struct PACKET_ZC_DIALOG_WINDOW_POS*)WFIFOP( fd, 0 );
+	p->PacketType = HEADER_ZC_DIALOG_WINDOW_POS;
+	p->x = x;
+	p->y = y;
+	WFIFOSET(fd, sizeof(struct PACKET_ZC_DIALOG_WINDOW_POS));
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:25230]:
```suggestion

	clif_send( &p, sizeof( p ), &sd.bl, SELF );
```

File: src/map/clif.cpp [hunk 33]
```diff
@@ -25187,6 +25187,65 @@ void clif_parse_reset_skill( int fd, map_session_data* sd ){
 #endif
 }
 
+void clif_sayDialogAlign(map_session_data *sd, int npcid, e_say_dialog_align align)
+{
+#if PACKETVER_MAIN_NUM >= 20210203 || PACKETVER_RE_NUM >= 20211103 || PACKETVER_ZERO_NUM >= 20221024
+	nullpo_retv(sd);
+
+	const int fd = sd->fd;
+	WFIFOHEAD(fd, sizeof(struct PACKET_ZC_DIALOG_TEXT_ALIGN));
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:25196]:
```suggestion
```

File: src/map/clif.cpp [hunk 34]
```diff
@@ -25187,6 +25187,65 @@ void clif_parse_reset_skill( int fd, map_session_data* sd ){
...
+	struct PACKET_ZC_DIALOG_TEXT_ALIGN *p = (struct PACKET_ZC_DIALOG_TEXT_ALIGN *)WFIFOP(fd, 0);
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:25197]:
```suggestion
	PACKET_ZC_DIALOG_TEXT_ALIGN p = {};

```

File: src/map/clif.cpp [hunk 35]
```diff
@@ -25187,6 +25187,65 @@ void clif_parse_reset_skill( int fd, map_session_data* sd ){
...
+	p->PacketType = HEADER_ZC_DIALOG_TEXT_ALIGN;
+	p->align = align;
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:25199]:
```suggestion
	p.PacketType = HEADER_ZC_DIALOG_TEXT_ALIGN;
	p.align = align;
```

File: src/map/clif.cpp [hunk 36]
```diff
@@ -25187,6 +25187,65 @@ void clif_parse_reset_skill( int fd, map_session_data* sd ){
...
+#endif  // PACKETVER_MAIN_NUM >= 20220504
+}
+
+void clif_set_npc_window_pos(map_session_data *sd, int x, int y)
+{
+#if PACKETVER_MAIN_NUM >= 20220504
+	nullpo_retv(sd);
+
+	const int fd = sd->fd;
+	WFIFOHEAD(fd, sizeof(struct PACKET_ZC_DIALOG_WINDOW_POS));
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:25225]:
```suggestion
```

File: src/map/clif.cpp [hunk 37]
```diff
@@ -25187,6 +25187,65 @@ void clif_parse_reset_skill( int fd, map_session_data* sd ){
...
+	struct PACKET_ZC_DIALOG_WINDOW_POS *p = (struct PACKET_ZC_DIALOG_WINDOW_POS*)WFIFOP( fd, 0 );
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:25226]:
```suggestion
	PACKET_ZC_DIALOG_WINDOW_POS p = {};

```

File: src/map/clif.cpp [hunk 38]
```diff
@@ -25187,6 +25187,65 @@ void clif_parse_reset_skill( int fd, map_session_data* sd ){
...
+	p->PacketType = HEADER_ZC_DIALOG_WINDOW_POS;
+	p->x = x;
+	p->y = y;
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:25229]:
```suggestion
	p.PacketType = HEADER_ZC_DIALOG_WINDOW_POS;
	p.x = x;
	p.y = y;
```

File: src/map/clif.cpp [hunk 39]
```diff
@@ -25187,6 +25187,65 @@ void clif_parse_reset_skill( int fd, map_session_data* sd ){
...
+#endif  // PACKETVER_MAIN_NUM >= 20220504
+}
+
+void clif_set_npc_window_pos_percent(map_session_data *sd, int x, int y)
+{
+#if PACKETVER_MAIN_NUM >= 20220504
+	nullpo_retv(sd);
+
+	const int fd = sd->fd;
+	WFIFOHEAD(fd, sizeof(struct PACKET_ZC_DIALOG_WINDOW_POS2));
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:25240]:
```suggestion
```

File: src/map/clif.cpp [hunk 40]
```diff
@@ -25187,6 +25187,65 @@ void clif_parse_reset_skill( int fd, map_session_data* sd ){
...
+	struct PACKET_ZC_DIALOG_WINDOW_POS2 *p = (struct PACKET_ZC_DIALOG_WINDOW_POS2*)WFIFOP( fd, 0 );
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:25241]:
```suggestion
	PACKET_ZC_DIALOG_WINDOW_POS2 p = {};

```

File: src/map/clif.cpp [hunk 41]
```diff
@@ -25187,6 +25187,65 @@ void clif_parse_reset_skill( int fd, map_session_data* sd ){
...
+	p->PacketType = HEADER_ZC_DIALOG_WINDOW_POS2;
+	p->x = x;
+	p->y = y;
```
Comment (Lemongrass3110) [inline src/map/clif.cpp:25244]:
```suggestion
	p.PacketType = HEADER_ZC_DIALOG_WINDOW_POS2;
	p.x = x;
	p.y = y;
```

File: doc/script_commands.txt [hunk 42]
```diff
@@ -11412,3 +11412,37 @@ Example:
...
 ---------------------------------------
+
+*setdialogalign(<align>);
+
+Set vertical or horizontal align in NPC dialog.
+Valid aligns:
+- horizontal align:
+  DIALOG_ALIGN_LEFT
+  DIALOG_ALIGN_CENTER
+  DIALOG_ALIGN_RIGHT
```
Comment (Lemongrass3110) [inline doc/script_commands.txt:11423]: @aleos89 @Atemo do we want HALIGN and VALIGN here? or is it okay like this?

File: src/map/script.cpp [hunk 43]
```diff
@@ -27123,6 +27123,54 @@ BUILDIN_FUNC(opentips){
 #endif
 }
 
+BUILDIN_FUNC(setdialogalign){
```
Comment (Lemongrass3110) [inline src/map/script.cpp:27126]: Applies to all script functions:
Would be nice to validate the input and report invalid usages.

**Review by Lemongrass3110 — CHANGES_REQUESTED** [review]

**Comment by ManyManaMan** [comment]
setdialogsize, setdialogpos, setdialogpospercent should be work on 20220504+

File: doc/script_commands.txt (hunk 42 above)
Comment (Pokye) [reply doc/script_commands.txt:11423]: And then decided? HALING/VALING or do I leave it as is?

File: src/map/script.cpp [hunk 44]
```diff
@@ -27123,6 +27123,54 @@ BUILDIN_FUNC(opentips){
...
+	map_session_data *sd;
+
+	if ( !script_rid2sd(sd) ) {
+		return SCRIPT_CMD_FAILURE;
+	}
+
+	clif_set_dialog_align(*sd, st->oid, (e_say_dialog_align)script_getnum(st, 2));
```
Comment (Lemongrass3110) [inline src/map/script.cpp:27133]:
```suggestion
	int32 align = script_getnum( st, 2 );
	
	if( align < DIALOG_ALIGN_LEFT || align > DIALOG_ALIGN_BOTTOM ){
		ShowError( "buildin_setdialogalign: Unknown align value %d\n", align );
		return SCRIPT_CMD_FAILURE;
	}

	clif_set_dialog_align( *sd, st->oid, static_cast<e_say_dialog_align>( align ) );
```

File: src/map/script.cpp [hunk 45]
```diff
@@ -27123,6 +27123,54 @@ BUILDIN_FUNC(opentips){
...
+
+BUILDIN_FUNC(setdialogsize){
+	map_session_data *sd;
+
+	if ( !script_rid2sd(sd) ) {
+		script_pushint(st, 0);
+		return SCRIPT_CMD_FAILURE;
+	}
+
+	clif_set_npc_window_size(*sd, script_getnum(st, 2), script_getnum(st, 3));
```
Comment (Lemongrass3110) [inline src/map/script.cpp:27146]:
```suggestion
	int32 x = script_getnum( st, 2 );
	
	if( x < 0 || x > INT16_MAX ){
		ShowError( "buildin_setdialogsize: x size %d is out of range [0,%d]\n", x, INT16_MAX );
		return SCRIPT_CMD_FAILURE;
	}
	
	int32 y = script_getnum( st, 3 );
	
	if( y < 0 || y > INT16_MAX ){
		ShowError( "buildin_setdialogsize: y size %d is out of range [0,%d]\n", y, INT16_MAX );
		return SCRIPT_CMD_FAILURE;
	}

	clif_set_npc_window_size( *sd, x, y );

```

File: doc/script_commands.txt [hunk 46]
```diff
@@ -11412,3 +11412,37 @@ Example:
...
+- vertical align:
+  DIALOG_ALIGN_TOP
+  DIALOG_ALIGN_MIDDLE
+  DIALOG_ALIGN_BOTTOM
+
+---------------------------------------
+
+*setdialogsize(<width>, <height>)
+
+Set size for NPC dialog.
```
Comment (Lemongrass3110) [inline doc/script_commands.txt:11434]:
```suggestion
Set size for NPC dialog in pixels.
```

File: src/map/script.cpp [hunk 47]
```diff
@@ -27123,6 +27123,54 @@ BUILDIN_FUNC(opentips){
...
+
+BUILDIN_FUNC(setdialogpos){
+	map_session_data *sd;
+
+	if ( !script_rid2sd(sd) ) {
+		script_pushint(st, 0);
+		return SCRIPT_CMD_FAILURE;
+	}
+
+	clif_set_npc_window_pos(*sd, script_getnum(st, 2), script_getnum(st, 3));
```
Comment (Lemongrass3110) [inline src/map/script.cpp:27158]:
```suggestion
	int32 x = script_getnum( st, 2 );
	
	if( x < 0 || x > INT16_MAX ){
		ShowError( "buildin_setdialogpos: x position %d is out of range [0,%d]\n", x, INT16_MAX );
		return SCRIPT_CMD_FAILURE;
	}
	
	int32 y = script_getnum( st, 3 );
	
	if( y < 0 || y > INT16_MAX ){
		ShowError( "buildin_setdialogpos: y position %d is out of range [0,%d]\n", y, INT16_MAX );
		return SCRIPT_CMD_FAILURE;
	}

	clif_set_npc_window_pos( *sd, x, y );

```

File: src/map/script.cpp [hunk 48]
```diff
@@ -27123,6 +27123,54 @@ BUILDIN_FUNC(opentips){
...
+
+BUILDIN_FUNC(setdialogpospercent){
+	map_session_data *sd;
+
+	if ( !script_rid2sd(sd) ) {
+		script_pushint(st, 0);
+		return SCRIPT_CMD_FAILURE;
+	}
+
+	clif_set_npc_window_pos_percent(*sd, script_getnum(st, 2), script_getnum(st, 3));
```
Comment (Lemongrass3110) [inline src/map/script.cpp:27170]:
```suggestion
	int32 x = script_getnum( st, 2 );
	
	if( x < 0 || x > 100 ){
		ShowError( "buildin_setdialogpospercent: x rate %d is out of range [0,100]\n", x );
		return SCRIPT_CMD_FAILURE;
	}
	
	int32 y = script_getnum( st, 3 );
	
	if( y < 0 || y > 100 ){
		ShowError( "buildin_setdialogpospercent: y rate %d is out of range [0,100]\n", y );
		return SCRIPT_CMD_FAILURE;
	}
	
	clif_set_npc_window_pos_percent( *sd, x, y );

```

**Review by Lemongrass3110 — CHANGES_REQUESTED** [review]

**Review by Lemongrass3110 — APPROVED** [review]

**Pull request merged**